lib/elixir/lib/calendar/datetime.ex

```diff
@@ -172,20 +172,213 @@ defmodule DateTime do
+
+  def from_naive(%{second: 60} = naive_datetime, "Etc/UTC", tz_db_or_config) do
+    {:ok, dt} = do_from_naive(naive_datetime, "Etc/UTC", 0, 0, "UTC")
+
+    case validate_positive_leap_second(dt, tz_db_or_config) do
+      :ok ->
+        {:ok, dt}
+
+      {:error, :outside_leap_second_data_range} ->
+        {:outside_leap_second_data_range, dt}
+
+      error ->
+        error
+    end
+  end
+
+  def from_naive(naive_datetime, "Etc/UTC", _) do
+    do_from_naive(naive_datetime, "Etc/UTC", 0, 0, "UTC")
+  end
+
+  def from_naive(%{calendar: Calendar.ISO} = naive_datetime, time_zone, tz_db_or_config) do
+    case TimeZoneDatabaseClient.time_zone_periods_from_wall_datetime(
+           naive_datetime,
+           time_zone,
+           tz_db_or_config
+         ) do
+      {:single, period} ->
+        do_from_naive_check_leap_second(
+          naive_datetime,
+          time_zone,
+          period.std_offset,
+          period.utc_offset,
+          period.zone_abbr,
+          tz_db_or_config
+        )
+
+      {:ambiguous, first_period, second_period} ->
+        {:ok, first_datetime} =
+          do_from_naive(
+            naive_datetime,
+            time_zone,
+            first_period.std_offset,
+            first_period.utc_offset,
+            first_period.zone_abbr
+          )
+
+        {:ok, second_datetime} =
+          do_from_naive(
+            naive_datetime,
+            time_zone,
+            second_period.std_offset,
+            second_period.utc_offset,
+            second_period.zone_abbr
+          )
+
+        {:ambiguous, first_datetime, second_datetime}
+
+      {:gap, {first_period, first_period_until_wall}, {second_period, second_period_from_wall}} ->
+        # `until_wall` is not valid, but any time just before is.
+        # So by subtracting a second and adding .999999 seconds
+        # we get the last microsecond just before.
+        before_naive =
+          first_period_until_wall
+          |> Map.put(:microsecond, {999_999, 6})
+          |> NaiveDateTime.add(-1)
+
+        after_naive = second_period_from_wall
+
+        {:ok, latest_datetime_before} =
+          do_from_naive(
+            before_naive,
+            time_zone,
+            first_period.std_offset,
+            first_period.utc_offset,
+            first_period.zone_abbr
+          )
+
+        {:ok, first_datetime_after} =
+          do_from_naive(
+            after_naive,
+            time_zone,
+            second_period.std_offset,
+            second_period.utc_offset,
+            second_period.zone_abbr
+          )
+
+        {:gap, latest_datetime_before, first_datetime_after}
+
+      {:error, _} = error ->
+        error
+    end
+  end
+
+  def from_naive(%{calendar: calendar} = naive_datetime, time_zone, tz_db_or_config)
+      when calendar != Calendar.ISO do
+    # For non-ISO calendars, convert to ISO, create ISO DateTime, and then
+    # convert to original calendar
```

Comment:
I want to mention this is not the most efficient nor the most correct way to go about this because it doesn't give the opportunity for databases which cannot be converted to ISO to handle timezones. The correct way would probably to add a new callback to `Calendar`. However, I am not even sure if there is a calendar that fits this criteria, so I would rather wait until an use case proves this necessary. So I am perfectly happy to ship this as is! :+1: :heart: